library/src/cache/patch_manager.rs

```diff
@@ -0,0 +1,883 @@
+
+    /// The highest patch number we have seen. This may be higher than the last booted
+    /// patch or next patch if we downloaded a patch that failed to boot.
+    highest_seen_patch_number: Option<usize>,
+}
+
+/// Abstracts the process of managing patches.
+#[cfg_attr(test, automock)]
+pub trait ManagePatches {
+    /// Copies the patch file at file_path to the manager's directory structure sets
+    /// this patch as the next patch to boot.
+    fn add_patch(&mut self, number: usize, file_path: &Path) -> Result<()>;
+
+    /// Returns the patch we most recently successfully booted from (usually the currently running patch),
+    /// or None if no patch is installed.
+    fn last_successfully_booted_patch(&self) -> Option<PatchInfo>;
+
+    /// Returns the next patch to boot, or None if:
+    /// - no patches have been downloaded
+    /// - the patch on disk is not bootable
+    fn next_boot_patch(&mut self) -> Option<PatchInfo>;
+
+    /// Records that the patch with number patch_number booted successfully and is
+    /// safe to use for future boots.
+    fn record_boot_success_for_patch(&mut self, patch_number: usize) -> Result<()>;
+
+    /// Records that the patch with number patch_number failed to boot, and ensures
+    /// that it will never be returned as the next boot or last booted patch.
+    fn record_boot_failure_for_patch(&mut self, patch_number: usize) -> Result<()>;
+
+    /// The highest patch number that has been added. This may be higher than the
+    /// last booted or next boot patch if we downloaded a patch that failed to boot.
+    fn highest_seen_patch_number(&self) -> Option<usize>;
+
+    /// Resets the patch manager to its initial state, removing all patches. This is
+    /// intended to be used when a new release version is installed.
+    fn reset(&mut self) -> Result<()>;
+}
+
+// This allows us to use the Debug trait on dyn ManagePatches, which is
+// required to have it as a property of UpdaterState.
+impl Debug for dyn ManagePatches {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "ManagePatches")
+    }
+}
+
+#[derive(Debug)]
+pub struct PatchManager {
+    /// The base directory used to store patch artifacts and state.
+    /// The directory structure created within this directory is:
+    ///  patches_state.json
+    ///  patches/
+    ///    <patch_number>/
+    ///      dlc.vmcode
+    ///    <patch_number>/
+    ///      dlc.vmcode
+    root_dir: PathBuf,
+
+    /// Metadata about the patches we have downloaded that is persisted to disk.
+    patches_state: PatchesState,
+}
+
+impl PatchManager {
+    /// Creates a new PatchManager with the given root directory. This directory is
+    /// assumed to exist. The PatchManager will use this directory to store its
+    /// state and patch binaries.
+    pub fn with_root_dir(root_dir: PathBuf) -> Self {
+        let patches_state = Self::load_patches_state(&root_dir).unwrap_or_default();
```

Comment:
Does the caller care that this might silently make a new one on failure?

Reply:
I guess it's not silent, you get an error! below.

Reply:
Yeah, we print the error, and there's not much we can do in the way of recovery if we can't deserialize the json file